File: Day 75/2487. Remove Nodes From Linked List.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:

    def reverse(self, head):

            prev = None
            curr = head

            while(curr is not None):
                forward = curr.next
                curr.next = prev
                prev = curr
                curr = forward
            
            return prev

    def removeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:

        head = self.reverse(head)

        temp = head

        maxi = temp.val

        while (temp.next is not None):
            
            curr = temp.next

            if(curr.val >= maxi):
                maxi = curr.val
                temp = temp.next
            
            else:
                temp.next = curr.next
                curr.next = None
                curr = None

        head = self.reverse(head)

        return head

        
                


            

File: Day 75/2487. Remove Nodes From Linked List.cpp
#include <iostream>
using namespace std;


  struct ListNode {
      int val;
      ListNode *next;
      ListNode() : val(0), next(nullptr) {}
      ListNode(int x) : val(x), next(nullptr) {}
      ListNode(int x, ListNode *next) : val(x), next(next) {}
  };

class Solution {
    private:

        ListNode* reverse(ListNode* head){

            ListNode* prev = nullptr;
            ListNode* curr = head;

            while(curr != nullptr){
                ListNode* forward = curr -> next;
                curr -> next = prev;
                prev = curr;
                curr = forward;
            }

            return prev;

        }

public:
    ListNode* removeNodes(ListNode* head) {
        
        head = reverse(head);

        ListNode* temp = head;

        int maxi = temp -> val;

        while(temp -> next != nullptr){
            
            ListNode* curr = temp -> next;

            if(curr -> val >= maxi){
                maxi = curr -> val;
                temp = temp -> next;
            }

            else{
                temp -> next = curr -> next;
                curr -> next = nullptr;
                delete(curr);
            }
        }

        head = reverse(head);

        return head;
        
        
    }
};

int main(){
  
  return 0;
}